node/bft/src/primary.rs

```diff
@@ -488,90 +488,99 @@ impl<N: Network> Primary<N> {
-                                }
-                            }
-                            // Check if the solution is still valid.
-                            if let Err(e) = self.ledger.check_solution_basic(solution_id, solution).await {
-                                trace!("Proposing - Skipping solution '{}' - {e}", fmt_id(solution_id));
-                                continue 'inner;
-                            }
+                    (TransmissionID::Transaction(transaction_id, checksum), Transmission::Transaction(transaction)) => {
+                        // Ensure the checksum matches. If not, skip the transaction.
+                        if !matches!(transaction.to_checksum::<N>(), Ok(transaction_checksum) if transaction_checksum == checksum )
+                        {
+                            trace!("Proposing - Skipping transaction '{}' - Checksum mismatch", fmt_id(transaction_id));
+                            continue;
                         }
-                        (
-                            TransmissionID::Transaction(transaction_id, checksum),
-                            Transmission::Transaction(transaction),
-                        ) => {
-                            // Ensure the checksum matches.
-                            match transaction.to_checksum::<N>() {
-                                Ok(transaction_checksum) if transaction_checksum == checksum => (),
+                        // Check if the transaction is still valid.
+                        // TODO: check if clone is cheap, otherwise fix.
+                        if let Err(e) = self.ledger.check_transaction_basic(transaction_id, transaction.clone()).await {
+                            trace!("Proposing - Skipping transaction '{}' - {e}", fmt_id(transaction_id));
+                            continue;
+                        }
+
+                        // Ensure the transaction doesn't bring the proposal above the spend limit.
+                        let block_height = self.ledger.latest_block_height() + 1;
+                        if N::CONSENSUS_VERSION(block_height)? >= ConsensusVersion::V4 {
+                            match self.ledger.compute_cost(transaction_id, transaction) {
+                                Ok(cost) if proposal_cost + cost <= N::BATCH_SPEND_LIMIT => proposal_cost += cost,
                                 _ => {
                                     trace!(
-                                        "Proposing - Skipping transaction '{}' - Checksum mismatch",
+                                        "Proposing - Skipping transaction '{}' - Batch spend limit surpassed",
                                         fmt_id(transaction_id)
                                     );
-                                    continue 'inner;
+
+                                    // Reinsert the transmission into the worker, O(n).
+                                    worker.shift_insert_front(id, transmission);
```

Comment:
Is this not an infinite loop?

1. The transmission exceeds the spend limit.
2. The transmission is added to the front of the worker
3. The worker pops the first item
4. Back to step 1.

Reply:
Might be smarter to pull other transactions that might fill the remaining space of the proposal, with smaller transactions. 

This also might be an opportune moment to finally implement the use of `priority_fee`, which was previously not factored into the ordering

Reply:
> If there is a huge transmission that is 99% of the batch spend limit, isn't your worker never going to create a proposal with a larger utilization?

Indeed, we don't currently attempt to fill the batch with smaller transmissions after the huge one is skipped. We could maintain _some_ degree of fairness by continuing to fill batch `n` until the spend limit or the batch size is reached and prioritise the large skipped transmission in batch `n + 1`. 

Not quite sure if new vectors are introduced with this approach or if it should be bundled with the priority fee changes but happy to make the adjustment here as well.